Multiple voting strategies

bench1ps opened the pull request:
A proposition of implementation regarding multiple voting strategies in order to allow for more complex toggles. I wrote an issue about it if you need some more information.

wjzijderveld commented:
Thanks for the PR, I hope to be able to take a closer look soon. Sorry it takes so long..
But feature wise it looks nice :)

wjzijderveld commented on lib/Qandidate/Toggle/Toggle.php:
```diff
@@ -61,10 +68,26 @@ public function activeFor(Context $context)
+    /**
+     * @param int $strategy
+     */
+    public function changeStrategy($strategy = self::STRATEGY_AFFIRMATIVE)
```
I don't see a reason why this argument would need a default value.

wjzijderveld commented on lib/Qandidate/Toggle/Toggle.php:
```diff
@@ -101,6 +135,21 @@ private function assertValidActiveStatus($status)
+     */
+    private function assertValidStrategy($strategy)
+    {
+        if ($strategy !== self::STRATEGY_AFFIRMATIVE && $strategy !== self::STRATEGY_CONSENSUS && $strategy !== self::STRATEGY_UNANIMOUS) {
```
I think using in_array here makes it a lot more readable:

```php
if (! in_array($strategy, array(
  self::STRATEGY_AFFIRMATIVE,
  self::STRATEGY_CONSENSUS,
  self::STRATEGY_UNANIMOUS
))) {
  // throw
}
```

wjzijderveld commented on examples/strategies.php:
```diff
@@ -0,0 +1,89 @@
+$context = new Context();
+$context->set('user_id', 41);
+$context->set('age', 23);
+var_dump($manager->active('toggling', $context)); // false
```
Could you add a newline at the end here?

bench1ps replied:
Done.

wjzijderveld commented:
@bench1ps Again, sorry for the lack of activity. 

Had some comments, but otherwise very nice!

bench1ps commented:
@wjzijderveld, thank you so much for your feedback.I fixed the issues you mentioned.

reenl commented:
I have no say in this, but I was actually working on something like this. Because issue #12 did not get mentioned from this PR.

However we have a different implementation. Since I already have it, I will also create a PR.

The main difference is I made the strategy part of the Toggle, not part of the activate function. The reasoning is that the strategy is bound to the conditions IMO, not to the "activity" of the toggle. Since Conditions are injected, I think the strategy should also be injected.

Other differences:
- I implemented consensus to fail at half
- There is no overlap between active and strategy values
- Strategy is only added to the serialization if not affirmative

bench1ps commented:
Hi @reenl, thank you for taking my issue into consideration. I apologize for the missing link between the issue I filed and my PR, this could have prevented our work from overlapping.

@wjzijderveld, what do you think ?

wjzijderveld commented:
@reenl @bench1ps 

Thanks for the work, both of you. I'll try to pay more attention to linking PR's to tickets as well. 

I like parts of both PR's, so that's going to complicate stuff maybe :smile: 

- I prefer the strategy as part of the constructor over the `activate` version
- I prefer to just always serialize the strategy

I don't really mind that the values of `strategy` overlapped with `status`, don't see how that would matter (unless when manually debugging the storage). Do you have a specific reason @reenl ?

reenl commented:
The strategy in my PR was conditionally serialized because of BC reasons, should be no problem to remove the if statement, some tests will need an update.

The overlap between `strategy` and `status` was to prevent the developer from accidentally mixing up the constants.

bench1ps commented:
Following the last few comments, you're both right in saying that it actually makes more sense for the strategy to be part of the toggle constructor. So I removed the `changeStrategy()` method and added a test to make sure that a toggle is never built using an unsupported strategy.

Regarding the behavior of the consensus strategy, is there any specific reason to make it fail at half @reenl ?

reenl commented:
There is no real reason. I created it because in democracy it would fail with half the votes, just a matter of taste. Only reason I mentioned it is to list changes nothing more :).

wjzijderveld commented:
@bench1ps I actually prefer the '>50%' version of concensus. The idea of consensus, is that there is an agreement, with exactly half there is no agreement, so it should fail imo.
But maybe it's a good sign that the name is a bit misleading, maybe we want to go with `majority`?

bench1ps commented:
@wjzijderveld To be honest I don't really have a preference, so I changed the strategy to fail at half because everybody agrees on this. And as a consequence I changed the strategy name to "majority" because it definitely makes more sense.

reenl commented:
:+1:

wjzijderveld commented:
I\m :+1: on this

@asm89 Could you take a look at it?

DaphDaOne commented:
I need that too ! \o/
Any idea on when it would be merged ? Thanks !

odolbeau commented:
:+1:

prcaen commented:
:+1:

dratel29 commented:
:+1:

wjzijderveld commented:
@bench1ps Would you mind squashing your commits?
I'd like to merge this today if possible :)

bench1ps commented:
@wjzijderveld, it's done :)

wjzijderveld commented:
Thanks @bench1ps for this!
@reenl Thanks to you too :smile:
